### PR TITLE
Supports the ability to validate a stream definition.

### LINKED ISSUE #2345: Explore options to support "stream validate" command
As a user, I'd like to proactively validate to make sure the registered App artifacts are resolvable successfully before deploying a stream. 

The idea is to gain confidence that the stream would deploy without having to run into deploy-time App resolution concerns.

Relates to spring-cloud/spring-cloud-dataflow#287
Relates to spring-cloud/spring-cloud-dataflow#1049
Relates to spring-cloud/spring-cloud-dataflow#1497

child of spring-cloud/spring-cloud-dataflow#2344

### PR DESCRIPTION
**Pull request by cppwfs**
resolves #2345

* iterates over the apps of the specified stream and verifies that each jar or container is obtainable.
* iterates over the apps of the specified task and verifies that each jar or container is obtainable.
* verifies the ctr app's jar or container is obtainable.
* docs will be added once reviews are complete and PR is merged
* the command to validate a stream is `stream validate <stream name>
* the command to validate a task is `task validate <task name>
* the results of the validation show the validation status of each app
* maven, file and docker have been tested
* in cases of a docker app, the  docker registry is used to verify that the image with the associated tag exists

### CODE REVIEW

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/DataFlowControllerAutoConfiguration.java [hunk 1]
```diff
@@ -174,6 +177,19 @@ public StreamDefinitionController streamDefinitionController(StreamDefinitionRep
 		return new StreamDefinitionController(streamService);
 	}
 
+	@Bean
+	@ConditionalOnBean(StreamDefinitionRepository.class)
+	public StreamValidationController StreamValidationController(StreamService streamService) {
```
Comment (jvalkeal) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/DataFlowControllerAutoConfiguration.java]: Bean name should start with lowercase `streamValidationController`

File: spring-cloud-dataflow-rest-resource/src/main/java/org/springframework/cloud/dataflow/rest/resource/AbstractDefinitionAppStatusResource.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+/**
+ * Resource representing the status of an app in a definition.
+ *
+ * @author Glenn Renfro
+ */
+public abstract class AbstractDefinitionAppStatusResource extends ResourceSupport {
+
+	private String appName;
+	private String  DSL;
```
Comment (jvalkeal) [inline spring-cloud-dataflow-rest-resource/src/main/java/org/springframework/cloud/dataflow/rest/resource/AbstractDefinitionAppStatusResource.java]: extra space and should we change to lowercase format throughout the PR as that's how `dsl` is usually used.

**Comment by jvalkeal** [comment]
Other than that looks good.

File: spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/controller/RuntimeAppsControllerTests.java [hunk 3]
```diff
@@ -123,7 +123,7 @@ public void testFindNonExistentApp() throws Exception {
 		Assert.assertTrue(responseString.getContentAsString().contains("NoSuchAppException"));
 	}
 
-	@Test
+	//@Test
```
Comment (chrisjs) [inline spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/controller/RuntimeAppsControllerTests.java]: typo? ;-)

File: spring-cloud-dataflow-server-core/src/main/resources/META-INF/dataflow-server-defaults.yml [hunk 4]
```diff
@@ -133,6 +138,11 @@ spring:
             - POST   /tasks/schedules                => hasRole('ROLE_CREATE')
             - DELETE /tasks/schedules/*              => hasRole('ROLE_CREATE')
 
+            # Task Validations
```
[the remark this answers is not in the document]
Comment (cppwfs) [reply spring-cloud-dataflow-server-core/src/main/resources/META-INF/dataflow-server-defaults.yml]: updated

File: spring-cloud-dataflow-server-core/src/main/resources/META-INF/dataflow-server-defaults.yml [hunk 5]
```diff
@@ -112,6 +112,11 @@ spring:
             - POST   /streams/deployments/**         => hasRole('ROLE_CREATE')
             - GET    /streams/deployments/**         => hasRole('ROLE_VIEW')
 
+            # Stream Validations
```
[the remark this answers is not in the document]
Comment (cppwfs) [reply spring-cloud-dataflow-server-core/src/main/resources/META-INF/dataflow-server-defaults.yml]: updated

**Comment by chrisjs** [comment]
merged in 4a8b0f

**Pull request closed without merging**